src/app/clusters/bindings/bindings.cpp

```diff
@@ -120,55 +163,127 @@ bool emberAfBindingClusterBindCallback(app::CommandHandler * commandObj, const a
+CHIP_ERROR BindingTableAccess::Read(FabricIndex fabricIndex, const ConcreteReadAttributePath & path,
+                                    AttributeValueEncoder & encoder)
+{
+    switch (path.mAttributeId)
     {
-        ChipLogError(Zcl, "Binding: Invalid request");
-        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_MALFORMED_COMMAND);
-        return true;
+    case BindingList::Id:
+        return ReadBindingTable(fabricIndex, path.mEndpointId, encoder);
+    default:
+        break;
     }
-    if (groupId)
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTableAccess::ReadBindingTable(FabricIndex fabricIndex, EndpointId endpoint, AttributeValueEncoder & encoder)
+{
+    DeviceLayer::AttributeList<Structs::BindingEntry::Type, EMBER_BINDING_TABLE_SIZE> bindingTable;
+
+    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
     {
-        bindingEntry = EmberBindingTableEntry::ForGroup(fabricIndex, groupId, localEndpoint, clusterId);
+        EmberBindingTableEntry entry;
+        emberGetBinding(i, &entry);
+        if (entry.type == EMBER_UNICAST_BINDING && entry.fabricIndex == fabricIndex)
+        {
+            Structs::BindingEntry::Type value = {
+                .nodeId     = entry.nodeId,
+                .groupId    = 0,
+                .endpointId = entry.remote,
+                .clusterId  = entry.clusterId,
+            };
+            bindingTable.add(value);
+        }
+        else if (entry.type == EMBER_MULTICAST_BINDING && entry.fabricIndex == fabricIndex)
+        {
+            Structs::BindingEntry::Type value = {
+                .nodeId     = 0,
+                .groupId    = entry.groupId,
+                .endpointId = 0,
+                .clusterId  = entry.clusterId,
+            };
+            bindingTable.add(value);
+        }
     }
-    else
+    return encoder.EncodeList([&bindingTable](const auto & subEncoder) {
+        for (auto & value : bindingTable)
+        {
+            ReturnErrorOnFailure(subEncoder.Encode(value));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR BindingTableAccess::Write(FabricIndex fabricIndex, const ConcreteDataAttributePath & path,
+                                     AttributeValueDecoder & decoder)
+{
+    switch (path.mAttributeId)
     {
-        bindingEntry = EmberBindingTableEntry::ForNode(fabricIndex, nodeId, localEndpoint, remoteEndpoint, clusterId);
+    case BindingList::Id:
+        return WriteBindingTable(fabricIndex, path.mEndpointId, decoder);
+    default:
+        break;
     }
+    return CHIP_NO_ERROR;
+}
 
-    uint8_t bindingIndex;
-    if (getBindingIndex(bindingEntry, &bindingIndex) != EMBER_SUCCESS)
+CHIP_ERROR BindingTableAccess::WriteBindingTable(FabricIndex fabricIndex, EndpointId endpoint, AttributeValueDecoder & decoder)
+{
+    BindingList::TypeInfo::DecodableType newBindingList;
+
+    ReturnErrorOnFailure(decoder.Decode(newBindingList));
+
+    // Add entries currently not in the binding table
```

Comment:
This is wrong.  Per spec, the order here matters, and this is not going to preserve the order of the incoming list...